include/Sample.h

```diff
@@ -0,0 +1,158 @@
+/*
+ * Sample.h - State for container-class SampleBuffer
+ *
+ * Copyright (c) 2023 saker <sakertooth@gmail.com>
+ *
+ * This file is part of LMMS - https://lmms.io
+ *
+ * This program is free software; you can redistribute it and/or
+ * modify it under the terms of the GNU General Public
+ * License as published by the Free Software Foundation; either
+ * version 2 of the License, or (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+ * General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public
+ * License along with this program (see COPYING); if not, write to the
+ * Free Software Foundation, Inc., 51 Franklin Street, Fifth Floor,
+ * Boston, MA 02110-1301 USA.
+ *
+ */
+
+#ifndef LMMS_SAMPLE_H
+#define LMMS_SAMPLE_H
+
+#include <cmath>
+#include <memory>
+
+#include "Note.h"
+#include "SampleBuffer.h"
+#include "lmms_export.h"
+
+#ifdef __MINGW32__
+#include <mingw.mutex.h>
+#include <mingw.shared_mutex.h>
+#else
+#include <mutex>
+#include <shared_mutex>
+#endif
+
+class QPainter;
+class QRect;
+
+namespace lmms {
+class LMMS_EXPORT Sample
+{
+public:
+	// values for buffer margins, used for various libsamplerate interpolation modes
+	// the array positions correspond to the converter_type parameter values in libsamplerate
+	// if there appears problems with playback on some interpolation mode, then the value for that mode
+	// may need to be higher - conversely, to optimize, some may work with lower values
+	static constexpr auto s_interpolationMargins = std::array<int, 5>{64, 64, 64, 4, 4};
+
+	enum class Loop
+	{
+		Off,
+		On,
+		PingPong
+	};
+
+	class LMMS_EXPORT PlaybackState
+	{
+	public:
+		PlaybackState(bool varyingPitch = false, int interpolationMode = SRC_LINEAR)
+			: m_varyingPitch(varyingPitch)
+			, m_interpolationMode(interpolationMode)
+		{
+			auto error = 0;
+			m_resampleState = src_new(m_interpolationMode, DEFAULT_CHANNELS, &error);
+			if (error != 0)
+			{
+				throw std::runtime_error{"Error creating resample state: " + std::string{src_strerror(error)}};
+			}
+		}
+
+		~PlaybackState() noexcept { src_delete(m_resampleState); }
```

Comment:
> A better approach, though more work, would be to extract the libsamplerate lifetime management into a new class, and go with the rule of zero in this class (no user-defined destructor needed).

Just to clarify, are you suggesting some static `Resampler` class? I had in that mind for other PRs, but I can do it here since its still relevant and applicable.